### ISSUE TITLE
Using C++ and QBDIPreload API on Android x86_64

### ISSUE THREAD

**Issue by pistach3** [QBDI/QBDI#185]
Hi,

I wanted to try the C++ `QBDI` API with `QBDIPreload` from the docs on Android.
For this I am on a `cuttlefish x86_64` emulator (Android `11`, API `30`).

The docs are not very clear for Android, so here are the steps I followed:

I started from the `fibonacci_cpp` example on the [master branch](https://github.com/QBDI/QBDI/tree/master/examples/cpp) and copied the `fibonacci.cpp` and `tracer_preload.cpp` files.

I use this `CMakeLists.txt` file:
```cmake
cmake_minimum_required(VERSION 3.2)
project(test)

add_executable(fibonacci_cpp "fibonacci.cpp")
set_target_properties(fibonacci_cpp PROPERTIES CXX_STANDARD 11 CXX_STANDARD_REQUIRED ON)

set(QBDI_PRELOAD "${CMAKE_CURRENT_SOURCE_DIR}/deps/libQBDIPreload.a")
set(QBDI_SHARED "${CMAKE_CURRENT_SOURCE_DIR}/deps/libQBDI.so")
set(QBDI_STATIC "${CMAKE_CURRENT_SOURCE_DIR}/deps/libQBDI.a")

include_directories("headers")

find_library(libc c)
find_library(libm m)
find_library(libz z)
target_link_libraries(fibonacci_cpp ${libc} ${QBDI_SHARED} ${tracer_preload_cpp})

add_library(tracer_preload_cpp SHARED tracer_preload.cpp)
target_link_libraries(tracer_preload_cpp ${libc} ${libm} ${libz} ${QBDI_STATIC} ${QBDI_PRELOAD}
```

The `deps` directory contains the `libQBDI*` files from the 0.8.0 release (`QBDI-0.8.0-android-X86_64.tar.gz`).

The `headers` directory contains all the files from `usr/local/include` from the same archive.

I build `fibonacci_cpp` and `libtracer_preload_cpp.so` without errors with:
```bash
cmake . -DCMAKE_TOOLCHAIN_FILE=<PATH_TO_NDK>/build/cmake/android.toolchain.cmake -DANDROID_ABI=x86_64 -DANDROID_NATIVE_API_LEVEL=30
make
```

I push `libQBDI.so` from the 0.8.0 release to `/system/lib64` on the device after remount. I also push `fibonacci_cpp` and `libtracer_preload_cpp.so` on the device, and also do a `setenforce 0`.

From here I do as per the docs:
```bash
# On the Android device
LD_PRELOAD=./libtracer_preload_cpp.so LD_BIND_NOW=1 ./fibonacci_cpp
```

However, even if I replace the content of the `onInstruction` function from `tracer_preload.cpp`, nothing different happens from simply running:
```bash
./fibonacci_cpp
```

Which gives:
```bash
Running fibonacci(1) ...
5b3584006b40:   push    rbp
5b3584006b41:   push    rbx
5b3584006b42:   push    rax
5b3584006b43:   mov     ebp, 1
5b3584006b48:   cmp     edi, 3
5b3584006b4b:   jl      40
5b3584006b75:   mov     eax, ebp
5b3584006b77:   add     rsp, 8
5b3584006b7b:   pop     rbx
5b3584006b7c:   pop     rbp
5b3584006b7d:   ret
fibonnaci(1) returns 1 after 0 recursions
```

Shouldn't I have additional output (or different output) by using `LD_PRELOAD`? Did I miss something? I don't really understand the appeal. Could you explain it to me a bit more or point me to a working Android example for `x86_64`?

Thank you in advance.

**Comment by nsurbay** [comment]
Hello,

In your example, you have two different usages of QBDI:

1. QBDI is compiled with the target code or the target code is imported (dlopen) inside QBDI process. The target code is then instrument by QBDI.
2. QBDI is injected with QBDIPreload in any process and instrument the main method.

With the command `LD_PRELOAD=./libtracer_preload_cpp.so LD_BIND_NOW=1 ./fibonacci_cpp`, you try to instrument (usage 2) the instrumentation (usage 1) of the function fibonnaci of `fibonacci_cpp`.

If you have the same result with and without the LD_PRELOAD, the library may not be preload or we have a bug in the QBDIPreload. Can you verify that your preload library is injected by printing the result of [getCurrentProcessMaps](https://qbdi.readthedocs.io/en/stable/api_cpp.html#_CPPv4N4QBDI21getCurrentProcessMapsEb) in the main method of `fibonacci_cpp` ?

```cpp
for (const QBDI::MemoryMap& m : QBDI::getCurrentProcessMaps(true)) {
    std::cout << std::hex << m.name << " [0x" << m.range.start() << ", 0x" << m.range.end() << ")" << std::endl;
}
```

**Comment by pistach3** [comment]
Hi again,

You are right, I think I am actually a bit confused by the `fibonacci_cpp` example :thinking:.

By the two use cases you explained, I gather that I could just preload the shared library `libtracer_preload_cpp.so` and use it on a random Android binary, which does not have anything to do with QBDI?

To come back to `fibonacci_cpp` and just to be sure to understand the situation well, I have the [onInstruction function in tracer_preload.cpp](https://github.com/QBDI/QBDI/blob/master/examples/cpp/tracer_preload.cpp#L5) which in my case only does:
```cpp
static QBDI::VMAction onInstruction(QBDI::VMInstanceRef vm, QBDI::GPRState *gprState, QBDI::FPRState *fprState, void *data)
{
    std::cout << "Hello tracer_preload onInstruction" << std::endl;
    return QBDI::CONTINUE;
}
```

So by using LD_PRELOAD and thanks to [the addCodeCB call in the `qbdipreload_on_run` method](https://github.com/QBDI/QBDI/blob/master/examples/cpp/tracer_preload.cpp#L30), I should have this callback executed for every instruction executed in `fibonacci_cpp`, right?


As asked, here is the result from `getCurrentProcessMaps` into the main method, where `libtracer_preload_cpp.so` is present:
```
/data/local/tmp/fibonacci_cpp [0x5d3183bc1000, 0x5d3183bd9000]
/data/local/tmp/fibonacci_cpp [0x5d3183bd9000, 0x5d3183c01000]
/data/local/tmp/fibonacci_cpp [0x5d3183c01000, 0x5d3183c04000]
/data/local/tmp/fibonacci_cpp [0x5d3183c04000, 0x5d3183c05000]
[anon:.bss] [0x5d3183c05000, 0x5d3183c07000]
 [0x73ebdd8c1000, 0x73ebdd8c2000]
[anon:scudo:secondary] [0x73ebdd8c2000, 0x73ebdd9c3000]
...
/system/lib64/libz.so [0x73ee6db58000, 0x73ee6db60000]
/system/lib64/libz.so [0x73ee6db60000, 0x73ee6db71000]
/system/lib64/libz.so [0x73ee6db71000, 0x73ee6db72000]
/system/lib64/libz.so [0x73ee6db72000, 0x73ee6db73000]
/apex/com.android.runtime/lib64/bionic/libc.so [0x73ee6db81000, 0x73ee6dbc3000]
/apex/com.android.runtime/lib64/bionic/libc.so [0x73ee6dbc3000, 0x73ee6dc4f000]
/apex/com.android.runtime/lib64/bionic/libc.so [0x73ee6dc4f000, 0x73ee6dc53000]
/apex/com.android.runtime/lib64/bionic/libc.so [0x73ee6dc53000, 0x73ee6dc56000]
[anon:.bss] [0x73ee6dc56000, 0x73ee6dc59000]
[anon:.bss] [0x73ee6dc59000, 0x73ee6dc5a000]
[anon:.bss] [0x73ee6dc5a000, 0x73ee6dc68000]
/apex/com.android.runtime/lib64/bionic/libdl.so [0x73ee6dc89000, 0x73ee6dc8a000]
/apex/com.android.runtime/lib64/bionic/libdl.so [0x73ee6dc8a000, 0x73ee6dc8b000]
/apex/com.android.runtime/lib64/bionic/libdl.so [0x73ee6dc8b000, 0x73ee6dc8c000]
 [0x73ee6dc8c000, 0x73ee6dc8d000]
[anon:.bss] [0x73ee6dc8d000, 0x73ee6dc8e000]
/data/local/tmp/libtracer_preload_cpp.so [0x73ee6dcc1000, 0x73ee6e105000]
/data/local/tmp/libtracer_preload_cpp.so [0x73ee6e105000, 0x73ee6e2f5000]
/data/local/tmp/libtracer_preload_cpp.so [0x73ee6e2f5000, 0x73ee6e3fb000]
/data/local/tmp/libtracer_preload_cpp.so [0x73ee6e3fb000, 0x73ee6e3fc000]
[anon:.bss] [0x73ee6e3fc000, 0x73ee6e400000]
/data/local/tmp/libQBDI.so [0x73ee6e400000, 0x73ee6e979000]
 [0x73ee6e979000, 0x73ee6e97a000]
/data/local/tmp/libQBDI.so [0x73ee6e97a000, 0x73ee6ea7b000]
/data/local/tmp/libQBDI.so [0x73ee6ea7b000, 0x73ee6ea7c000]
[anon:.bss] [0x73ee6ea7c000, 0x73ee6ea7e000]
/apex/com.android.runtime/lib64/bionic/libm.so [0x73ee6ea87000, 0x73ee6eaaa000]
/apex/com.android.runtime/lib64/bionic/libm.so [0x73ee6eaaa000, 0x73ee6ead3000]
/apex/com.android.runtime/lib64/bionic/libm.so [0x73ee6ead3000, 0x73ee6ead5000]
/apex/com.android.runtime/lib64/bionic/libm.so [0x73ee6ead5000, 0x73ee6ead6000]
...
```

Actually, by re-reading the `QBDIPreload API` docs:
> It currently only works under Linux using the LD_PRELOAD mechanism and macOS using the DYLD_INSERT_LIBRARIES mechanism. For other platforms please check out Frida/QBDI API instead.

(I guess this means no Android?)

Thanks!

**Comment by nsurbay** [comment]
We have not try to run QBDIPreload on Android but the LD_PRELOAD mechanism seems to work as `/data/local/tmp/libtracer_preload_cpp.so` is mapped in the process. We may need a custom version of QBDIPreload for Android.

> By the two use cases you explained, I gather that I could just preload the shared library libtracer_preload_cpp.so and use it on a random Android binary, which does not have anything to do with QBDI?

Yes. When you compile your qbdipreload library with your callback (include in `tracer_preload.cpp`), QBDI is statically linked inside. You don't need to place `libQBDI.so` on your system. The library should intercept the symbol `__libc_start_main` and later, the main method of the executable. When the main method try to run, your method `qbdipreload_on_run` will be called. You just need to add your callback and run the VM to instrument the main method of the program.

In your case, I think the method  `qbdipreload_on_run` is never called because the interception doesn't work. I need to investigate the issue.

> So by using LD_PRELOAD and thanks to the addCodeCB call in the qbdipreload_on_run method, I should have this callback executed for every instruction executed in fibonacci_cpp, right?

Yes, if you test the same things on Linux (where QBDIPreload should work), the callback will be called for each instruction executed in the binary (`fibonacci_cpp` or `/bin/ls` or any other dynamic binary).


To summarize the usage of QBDIPreload : 

- If you can compile the target code or the target code is inside a library, you should not used QBDIPreload and adapt `fibonacci.cpp` by replacing the fibonacci fonction by your target.
- If the target is a closed-source dynamic executable, you can used QBDIPreload to instrument it without modified the binary.

As QBDIPreload seems to have issue on Android, you may used frida to inject yourself, or modify the target binary into a library with the main method exported (for example, with [LIEF](https://lief.quarkslab.com/doc/latest/tutorials/08_elf_bin2lib.html)  ).

**Comment by pistach3** [comment]
Thank you for all the explanations and clearing up my confusion 😄!

I will definitely have a deeper look at LIEF for Android and watch for QBDIPreload updates for this platform.
